staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/defaults.go

```diff
@@ -71,4 +74,20 @@ func SetDefaults_CustomResourceDefinitionSpec(obj *CustomResourceDefinitionSpec)
 			{Name: "Age", Type: "date", Description: swaggerMetadataDescriptions["creationTimestamp"], JSONPath: ".metadata.creationTimestamp"},
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceWebhookConversion) {
```

Comment:
Different defaulting depending on a gate feels wrong.

The more I think about this whole PR, the more I think we should add the new fields per version independent of the feature gate.

Another concern is that with this defaulting we have **a breaking API change**:
- create a CRD with multiple versions and some openapi schema.
- the defaulting takes place.
- update the CRD by changing the schema => only the first version schema is changed (while in 1.11 all schemata were changed).

/cc @liggitt @deads2k what's your take on this?

@kubernetes/api-reviewers

Reply:
defaulting a field based on another field is the same issue we ran into with selector defaulting in some of the workload apis... defaulting made the initial value sticky in other parts of the object, and didn't update when the source field changed in later updates.

for those, in later releases, we removed the defaulting because it was problematic

Reply:
Spent time with @erictune on this and what we came up with is:
We don't do defaulting. I agree that will be a problem now and in future. However by not doing defaulting we will have forward incompatibility. When on 1.11 we have a CRD with multiple versions and we upgrade to 1.12, all of these fields will be disabled for all versions except first version (because they are nil and not defaulted. FeatureGate does not matter here, if we have feature gate just g/1.12/13/).
To fix that we came up with an idea of having an actual flag in the API to enable the behaviour explicitly and by accepting some limitation we think we should use Spec.Conversion.Strategy for that not a new field. It is uncommon to have multiple schemas with no conversion but the limitation we accept here is to have multiple subresource and additionalprintcolumn setting per version only when Conversion is not None (e.g. Webhook).

So to sum it up, we don't default and we gate all behaviour using Spec.Conversion.Strategy field which is set by default to None.

What do you guys think?

Reply:
> update the CRD by changing the schema

For PUT requests the defaulting takes place and changes the per-version schema. For patch requests only the first version schema is changed. After we introduce multi-version schema, it's hard to tell if a user wants to change all schema or only the first version schema, so it is a breaking change. 

> defaulting made the initial value sticky in other parts of the object, and didn't update when the source field changed in later updates.

That's true. But without defaulting, the versions other than the top-level version loss schema when we enable the feature gate, which is also a breaking change.

Reply:
> For PUT requests the defaulting takes place and changes the per-version schema. For patch requests only the first version schema is changed.

I'm not following this... updates from put and patch requests are not defaulted or converted differently

Reply:
According to my reading of the [API compatibility guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#backward-compatibility-gotchas), the problem started when we introduced `versions` instead of `additionalVersions`. The document basically asserts that there is no way to get out of this without breaking compatibility in some way.

If we accept that, our task here is to choose which breaking change to make. Just to play devil's advocate: what if we choose to break people in a really obvious way and say, "oops! we messed up. `versions` should have been `additionalVersions` and we've fixed it now but you'll have to update your objects if you used the new field already," instead of trying to minimize the scope of the inevitable breakage by adding more complexity.

After all, isn't this why we still call CRD beta? If we do this, we'd be back on track to have a clear path to migrate from `version+additionalVersions` to just `versions` in `v1`. If we don't, the eventual migration to `v1` will remain muddied ("this field gets put over there during conversion, but only if this other field is set to that value...").

Reply:
@liggitt I was thinking if a user sends a put request with updated top-level schema and nil per-version schema, in which case the defaulting will take place and update per-version schema, but actually patch requests can achieve the same, and I suppose this isn't how update requests are used in reality.

Reply:
I like `additionalVersions` as @enisoc suggests. It would remove a lot of the ugly special cases around `versions[0]`. But this is a pretty hard beta API breaking change.

About @mbohlool and @erictune's "we don't default" discussion: we could logically default, but don't actually set the fields. I.e. if additionalColumns is nil, we use the global additionalColumns, same for subresources and schema.

Reply:
> we could logically default, but don't actually set the fields. I.e. if additionalColumns is nil, we use the global additionalColumns, same for subresources and schema.

that makes sense to me, especially for fields that in many cases are likely to be identical across versions. that would mean you'd need to keep the top level field in v1, and keep the "shared top-level field overridden by version-specific field" behavior.

this is similar to the pod/container security context handling, where you can set options at the pod level or for a particular container

Reply:
> I like additionalVersions as @enisoc suggests. It would remove a lot of the ugly special cases around versions[0]

I think I actually prefer the top-level-field with optional per-version override approach

Reply:
override approach is not quite what we propose. Proposal is:
- if Spec.Conversion.Strategy=None, Per version fields are not supported, they cannot be set and validation will fail on non-nil values for per-version fields.
- If Spec.Conversion.Strategy!=None then per version fields will be honored. No defaulting will happened, top level fields will be used for first version and validation will make sure users cannot set fields for versions[0].

Because we do not do any defaulting, @liggitt concern on defaulting using other fields is addressed. Also no defaulting means we do not have any problem with your concern @enisoc, right?

Reply:
In other words:

1. What @mbohlool proposes above allows us to move to `additionalVersions` in v1.
2. In contrast, with the variant implementing "top-level fields are used as global default if version fields are unset" would lead us to a `versions` field in v1 and removal of the global `version` field in v1.

Which direction do we want to go?

Reply:
I'd hoped the API we ended up with in v1 would treat all the versions consistently. what are the downsides to optional per-version settings (for all versions, including the first one), falling through to the top-level fields?

Reply:
Agree we'd like to move to single list in v1.
I like "new fields per version independent of the feature gate".

Reply:
What I suggested would also lead us to versions field in v1. It does not need the nil means default thing. Nil means none/disable as it is right now for global.

Reply:
I am rooting for Strategy=none gating solution explained above not the field=nil means default to top level solution. I think nil should still mean disable and we should be able to set that per version and it fits well with crd v1's unified versions list.

Reply:
@mbohlool variant 1 is not uniform for all strategies. How should that look like in v1?

I tend to agree with @liggitt and @erictune that variant 2 is cleaner. It avoids strange semantics for different strategies and lets us to allow multiple schemata+columns+subresources without the gate.

Reply:
For obvious reasons, we should probaby restrict multi schemata with the gate as this does not make sense without webhook conversion.

Reply:
From our offline discussion (with @sttts) I agree with the following:
- No defaulting
- nil means use the top level fields
I think we should add this:
- Fields of version[0] should be nil
and my reasoning is the [API compatibility](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#backward-compatibility-gotchas) guidelines document:
```
Old clients must continue to function properly while only manipulating the old field. New clients must be able to function properly while only manipulating the new field.
```
If a new client sets `additionalprintcolumns` for version[0] it overrides global. if an old client patches global additonalprintcolumns which it thinks correspond to version[0], it won't see the changes hence it won't function properly. With the above addition we keep backward compatibility according to our docs.

Reply:
> Fields of version[0] should be nil

That won't round trip with v1 if we don't want to have the same restriction in v1 (which I hope we don't)

I don't see the issue with allowing overriding for the first version as well

Reply:
This may (without designing v1 now) make v1 API ugly to make it round trippable to v1beta1 but I think we should keep backward compatibility no matter how ugly.

Reply:
Allowing override for version[0] is backward incompatible according to out guidelines

Reply:
I am still catching up on this issue but a conversation Tim & I had may be relevant--please read it: https://groups.google.com/forum/#!msg/kubernetes-sig-architecture/4Ibjx2pn1aU/la81KWmcDwAJ

Reply:
If we do the logical defaulting, when converting from v1 to v1beta1 we won't be able to set value for top-level fields because we don't know which per-version fields in v1 is the "global" value in v1beta1. That means if we perform a roundtrip conversion from an v1beta1 object to v1, and back to v1beta1, the object will be different.

Reply:
> If we do the logical defaulting, when converting from v1 to v1beta1 we won't be able to set value for top-level fields because we don't know which per-version fields in v1 is the "global" value in v1beta1

Keeping the top-level and per-version fields in v1, with no API-level defaulting or copying of values between them in conversion, keeps round-tripping between the two versions very straightforward, and I think gives us a v1 API that is reasonably shaped.

I think it is extremely unlikely that two actors are going to define different parts of the same CRD independently, using different methods for specifying validation, without coordination. Someone who does have that use case is free to leave version[0] fields unset in the newer client and use only the top-level fields for that version, letting the older uncoordinated client set/reset those top-level fields.

Requiring that coordination in v1beta1 seems like a reasonable tradeoff to get a more reasonable long-term API.

Reply:
> Keeping the top-level and per-version fields in v1, with no API-level defaulting or copying of values between them in conversion

If we keep both the top-level schema and `versions[0].schema` in v1, which one should the apiserver honor? I'm assuming a client may choose to patch either of them.

Reply:
I'd expect the semantics to be the same as v1beta1 - the version-specific field would rule if specified

Reply:
@liggitt Do you suggest that we keep the logical defaulting in v1?

Reply:
Not arguing on the likelihood of the usecase yet but are you saying "keep backward compatibility unless the usecase breaking it is too narrow" should be in our policy doc? Let's agree or disagree on that first.

Reply:
I'm still wondering if in v1 API we could have a single `versions` field and remove top-level fields:

1. if in v1beta1 we allow `versions[0]` fields to be non-nil to override global value, we cannot roundtrip conversion because from v1beta1 to v1 one of {top-level schema, `versions[0].schema`} will get dropped.

2. if in v1beta1 we require `versions[0]` fields to be nil, we can roundtrip conversion between v1beta1 and v1 by defaulting (copying values). 

What's the downside of having a single `versions` list in v1?

Reply:
Given these facts:
- The difference between solutions we discuss right now is the limitations for version[0]. What we do in v1 can be the same regardless of this limitation (e.g. keeping top level fields, allowing overrides even for version 0). That is yet to be discussed.
- The Approved KEP proposed we do not set these values for version[0] hence we have an approved design that allow us to limit version[0]. I am not suggesting we should stop discussing it, but this is a fact.

I suggest:
- Let's get this unblocked by limiting overrides for version[0] for now and keep discussing.
- If we decide we do not want limitations for version[0], that we can remove with no backward compatibility issue. If we passed code freeze, that would be an easier exception to get instead of this whole feature.
- If we decided we do want to keep the limitation, then we can discuss v1 API. There are limitations in round-tripping to single versions field but that can be discussed.

Reply:
> If we decide we do not want limitations for version[0], that we can remove with no backward compatibility issue

Sounds good to me